document-store/src/main/java/org/hypertrace/core/documentstore/expression/impl/ArrayRelationalFilterExpression.java

```diff
@@ -0,0 +1,56 @@
+package org.hypertrace.core.documentstore.expression.impl;
+
+import com.google.common.base.Preconditions;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Value;
+import org.hypertrace.core.documentstore.expression.operators.ArrayOperator;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.parser.FilterTypeExpressionVisitor;
+
+/**
+ * Expression representing a condition for filtering on array fields
+ *
+ * <p>Example: If color is an array field <code>
+ * ANY(color) IN ('Blue', 'Green')
+ * </code> can be constructed as <code>
+ *   ArrayRelationalFilterExpression.builder()
+ *    .operator(ANY)
+ *    .filter(
+ *      RelationalExpression.of(
+ *        IdentifierExpression.of("color"),
+ *        IN,
+ *        ConstantExpression.ofStrings("Blue", "Green")
+ *    )
+ *    .build();
+ * </code>
+ */
+@Value
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class ArrayRelationalFilterExpression implements FilterTypeExpression {
```

Comment:
Shouldn't the DocumentArrayFilterExpression suffice? Also DocumentArrayFilterExpression explicitly mark the array source. Here, we will have to deduce (or implicit assumption).

Reply:
> Shouldn't the DocumentArrayFilterExpression suffice?

We would want to treat primitive arrays and document (object) arrays differently, because
1. In the case of primitive arrays, we do not have the LHS and have a special parsing mechanism
2. In the case of document arrays, we always need the LHS and its path is relative to the array source

> Also DocumentArrayFilterExpression explicitly mark the array source. Here, we will have to deduce (or implicit assumption).

Yes. That's by design. For primitive arrays, the LHS itself is the array source (so we don't need to explicitly ask the same from the users again).

Reply:
For primitive types: would name `ArrayFilterExpression` suffice?
For non-primitive : `DocumentArrayFilterExpression`?

Reply:
So, I am assuming that naming is to do with that we just want to allow only relation expression in primitive array types.

Reply:
> So, I am assuming that naming is to do with that we just want to allow only relation expression in primitive array types.

Yes, exactly. `DocumentArrayFilterExpression` is an `Expression` to `Filter` `Arrays` containing `Document`s. So, the `Filter` in this context means any filter (could be relational, logical or even another array filter).

`ArrayRelationalFilterExpression` is an `Expression` to apply `RelationalFilter` on `Arrays`. In this context, it is evident that only relational filtering can be applied using this. Deliberately left out the term `Primitive` here, because one could also do an equality comparison of 2D arrays using this filter. (Something like `ANY(2D_array) = Document.of("['hello', 'world']")`)